File: parse_into_db.py
import os
os.environ["HF_HUB_OFFLINE"] = "1"

from qdrant_client import QdrantClient
from qdrant_client.models import VectorParams, Distance, PointStruct
from sentence_transformers import SentenceTransformer
from langchain_text_splitters import RecursiveCharacterTextSplitter


class ParseIntoDB:

    def init_db(self, delete_collection):
        self.db = QdrantClient(url="http://localhost:6333")
        self.model = SentenceTransformer("paraphrase-multilingual-MiniLM-L12-v2")  # 384 dimensioni, leggero
        self.COLLECTION = "lapiazza"
        
        if not self.db.collection_exists(self.COLLECTION):
            self.db.create_collection(
                collection_name=self.COLLECTION,
                vectors_config=VectorParams(size=384, distance=Distance.COSINE)
            )
        else:
            if delete_collection:
                self.db.delete_collection(collection_name=self.COLLECTION)            
            self.db.create_collection(
                collection_name=self.COLLECTION,
                vectors_config=VectorParams(size=384, distance=Distance.COSINE)
            )

    def __init__(self, chunk_size=800, chunk_overlap=100, vector_size=384, delete_collection=False):
        self.chunk_size = chunk_size
        self.chunk_overlap = chunk_overlap
        self.vector_size = vector_size
        self.init_db(delete_collection=delete_collection)

    def add(self, document):
        splitter = RecursiveCharacterTextSplitter(
            chunk_size=self.chunk_size, 
            chunk_overlap=self.chunk_overlap,
            separators=["\n\n", "\n", ". ", " ", ""] 
        )
        documents = splitter.split_text(document)
        # print(len(documents))
        # for i, d in enumerate(documents):
        #     print(i,len(d),d)
        #     print()

        embeddings = self.model.encode(documents)
        points = [
            PointStruct(id=i, vector=emb.tolist(), payload={"text": doc})
            for i, (emb, doc) in enumerate(zip(embeddings, documents))
        ]
        self.db.upsert(collection_name=self.COLLECTION, points=points)
        print("Inseriti", len(points), "documenti")

    def search(self, query):
        query_vector = self.model.encode(query).tolist()
        results = self.db.query_points(
            collection_name=self.COLLECTION,
            query=query_vector,
            limit=3,
            with_payload=True
        ).points

        print(f"\nRisultati per: '{query}'\n")
        for r in results:
            print(f"score={r.score:.4f} | id={r.id} | text={r.payload['text']}\n")

if __name__ == "__main__":
    p = ParseIntoDB(chunk_size=800, chunk_overlap=50, vector_size=384, delete_collection=True)
    with open("article1.txt", "r") as f:
        text = f.read()
    p.add(text)
    p.search("Parlami di un autore di quadri")


# // List all collections
# GET collections

# // Get collection info
# GET collections/test_docs

# // List points in a collection, using filter
# POST /collections/test_docs/points/scroll
# {
#   "limit": 10,
#   "with_payload": true,
#   "with_vector": false
# }

File: node/parse_into_db.js
import { pipeline } from "@huggingface/transformers";
import { QdrantClient } from "@qdrant/js-client-rest";
import { RecursiveCharacterTextSplitter } from "@langchain/textsplitters";
import fs from "fs";

class ParseIntoDB {
  constructor({ chunkSize = 800, chunkOverlap = 100, vectorSize = 384 } = {}) {
    this.chunkSize = chunkSize;
    this.chunkOverlap = chunkOverlap;
    this.vectorSize = vectorSize;
    this.COLLECTION = "lapiazza";
  }

  async init(deleteCollection = false) {
    this.db = new QdrantClient({ url: "http://localhost:6333" });
    this.model = await pipeline("feature-extraction", "Xenova/paraphrase-multilingual-MiniLM-L12-v2");

    const { exists } = await this.db.collectionExists(this.COLLECTION);
    if (exists && deleteCollection) {
      await this.db.deleteCollection(this.COLLECTION);
    }
    if (!exists || deleteCollection) {
      await this.db.createCollection(this.COLLECTION, {
        vectors: { size: this.vectorSize, distance: "Cosine" },
      });
    }
  }

  async encode(texts) {
    const output = await this.model(texts, { pooling: "mean", normalize: true });
    return output.tolist();
  }

  async add(document) {
    const splitter = new RecursiveCharacterTextSplitter({
      chunkSize: this.chunkSize,
      chunkOverlap: this.chunkOverlap,
      separators: ["\n\n", "\n", ". ", " ", ""],
    });
    const documents = await splitter.splitText(document);

    console.log(documents.length);

    const embeddings = await this.encode(documents);
    const points = documents.map((doc, i) => ({
      id: i,
      vector: embeddings[i],
      payload: { text: doc },
    }));

    await this.db.upsert(this.COLLECTION, { points });
    console.log("Inseriti", points.length, "documenti");
  }

  async search(query) {
    const [queryVector] = await this.encode([query]);
    const results = await this.db.query(this.COLLECTION, {
      query: queryVector,
      limit: 3,
      with_payload: true,
    });

    console.log(`\nRisultati per: '${query}'\n`);
    for (const r of results.points) {
      console.log(`score=${r.score.toFixed(4)} | id=${r.id} | text=${r.payload.text}\n`);
    }
  }
}

async function main() {
  const p = new ParseIntoDB({ chunkSize: 800, chunkOverlap: 50, vectorSize: 384 });
  await p.init(true);
  const text = fs.readFileSync("../article1.txt", "utf-8");
  await p.add(text);
  await p.search("Parlami di un autore di quadri");
}

main();
